src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -307,6 +351,62 @@ def __init__(self, target: Text, options: Optional[ChannelArgumentType],
         self._loop = asyncio.get_event_loop()
         self._channel = cygrpc.AioChannel(_common.encode(target), options,
                                           credentials, self._loop)
+        self._ongoing_calls = _OngoingCalls()
+
+    async def __aenter__(self):
+        """Starts an asynchronous context manager.
+
+        Returns:
+          Channel the channel that was instantiated.
+        """
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        """Finishes the asynchronous context manager by closing the channel.
+
+        Still active RPCs will be cancelled.
+        """
+        await self._close(None)
+
+    async def _close(self, grace):
+        if self._channel.closed():
+            return
+
+        # No new calls will be accepted by the Cython channel.
+        self._channel.closing()
+
+        if grace:
+            # pylint: disable=unused-variable
+            _, pending = await asyncio.wait(self._ongoing_calls.calls,
+                                            timeout=grace,
+                                            loop=self._loop)
+
+            if not pending:
+                return
+
+        # A new set is created acting as a shallow copy because
+        # when cancellation happens the calls are automatically
+        # removed from the originally set.
+        calls = WeakSet(data=self._ongoing_calls.calls)
+        for call in calls:
+            call.cancel()
```

Comment:
Do we need to worry that `call` got deallocated already?

Reply:
as it is right now yes, otherwise the iterator would have unpredictable behavior since each call to `call.cancel` will end up modifying the `WeakSet`.

Reply:
I haven't experiment much about `WeakRef` with GC behaviors. Will the `close` code path be ran as a part of `__del__`? If so, there might not be a reliable guarantee that all `Call` objects are still intact. My question is about if the referenced `Call` object in the `WeakSet` got deallocated, maybe it will generate a weird exception? Should we add a if condition to check if `call` is not `None`?

Reply:
> My question is about if the referenced Call object in the WeakSet got deallocated, maybe it will generate a weird exception? Should we add a if condition to check if call is not None?

In case of being deallocated won't be returned in the iteration, `None` objects are filtered automatically by `WeaKSet` [1].

[1] https://github.com/python/cpython/blob/master/Lib/_weakrefset.py#L62

Reply:
Sounds good!